Sources/NIOHTTP1/HTTPServerPipelineHandler.swift

```diff
@@ -0,0 +1,159 @@
+                // response.
+                self = .responseEndPending
+            case .responseEndPending:
+                preconditionFailure("Received second request")
+            }
+        }
+    }
+
+    /// The connection state
+    private var state = ConnectionState.idle
+
+    /// While we're waiting to send the response we don't read from the socket.
+    /// This keeps track of whether we need to call read() when we've send our response.
+    private var readPending = false
+
+    /// The buffered HTTP requests that are not going to be addressed yet. In general clients
+    /// don't pipeline, so this initially allocates no space for data at all. Clients that
+    /// do pipeline will cause dynamic resizing of the buffer, which is generally acceptable.
+    private var requestBuffer = CircularBuffer<NIOAny>(initialRingCapacity: 0)
+
+    public func channelRead(ctx: ChannelHandlerContext, data: NIOAny) {
+        if case .responseEndPending = self.state {
+            self.requestBuffer.append(data)
+            return
+        }
+
+        if case .end = self.unwrapInboundIn(data) {
+            // New request is complete. We don't want any more data from now on.
+            self.state.requestEndReceived()
+        }
+        ctx.fireChannelRead(data)
+    }
+
+    public func write(ctx: ChannelHandlerContext, data: NIOAny, promise: EventLoopPromise<Void>?) {
+        assert(self.state != .requestEndPending, "Unexpected data flow")
+        var startReadingAgain = false
+        if case .end = self.unwrapOutboundIn(data) {
+            startReadingAgain = true
+        }
+
+        ctx.write(data, promise: promise)
+
+        if startReadingAgain {
+            self.state.responseEndReceived()
+            self.deliverPendingRequests(ctx: ctx)
+            self.startReading(ctx: ctx)
+        }
+    }
+
+    public func read(ctx: ChannelHandlerContext) {
+        if case .responseEndPending = self.state {
+            self.readPending = true
+        } else {
+            ctx.read()
+        }
+    }
+
+    /// A response has been sent: we can now start passing reads through
+    /// again, and send any read() call we may have swallowed.
+    private func startReading(ctx: ChannelHandlerContext) {
+        if self.readPending {
+            self.readPending = false
+            ctx.read()
+        }
+    }
+
+    /// A response has been sent: deliver all pending requests and
+    /// mark the channel ready to handle more requests.
+    private func deliverPendingRequests(ctx: ChannelHandlerContext) {
+        var deliveredRead = false
+
+        while self.state != .responseEndPending, let read = self.requestBuffer.first {
+            _ = self.requestBuffer.removeFirst()
+            self.channelRead(ctx: ctx, data: read)
+            deliveredRead = true
+        }
+
+        if deliveredRead {
+            ctx.fireChannelReadComplete()
+        }
+    }
```

Comment:
@Lukasa do we need to do something "smart" for the buffered writes if `close(ctx, mode, promise)` is called ? Maybe its ok to just drop these on the floor tho

Reply:
I don't think so. If we got a close there's not really anything smart we can do with them anyway.

Reply:
sounds good...